src/App.tsx

```diff
@@ -1,223 +1,243 @@
+import { ErrorNotification } from './components/ErrorNotification/ErrorNotification';
+
+import {
+  createTodo,
+  deleteTodoById,
+  getTodos,
+  updateTodoById,
+} from './api/todos';
+
+import { filterTodos } from './helpers/completedFilter';
+
+import { Todo } from './types/Todo';
+import { FilterType } from './types/FilterType';
 
 export const App: React.FC = () => {
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
   const newTodoField = useRef<HTMLInputElement>(null);
 
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [complitedFilter, setComplitedFilter] = useState(FilterType.ALL);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isNewTodoLoading, setIsNewTodoLoading] = useState(false);
+  const [updatingTodos, setUpdatingTodos] = useState<number[]>([]);
+
   useEffect(() => {
     // focus the element with `ref={newTodoField}`
     if (newTodoField.current) {
       newTodoField.current.focus();
     }
+
+    if (user) {
+      getTodos(user.id)
+        .then(setTodos)
+        .catch(() => {
+          setErrorMessage('Unable to download todos');
+        });
+    }
+  }, []);
+
+  const uncompletedTodosLength = useMemo(() => (
+    todos.filter(todo => !todo.completed).length
+  ), [todos]);
+
+  const completedTodosLength = useMemo(() => (
+    todos.filter(todo => todo.completed).length
+  ), [todos]);
+
+  const visibleTodos = useMemo(() => {
+    return filterTodos(todos, complitedFilter);
+  }, [complitedFilter, todos]);
+
+  const deleteTodo = useCallback(async (todoId: number) => {
+    try {
+      const deleteResponse = await deleteTodoById(todoId);
+
+      setTodos(prevTodos => (
+        prevTodos.filter(todo => todo.id !== todoId)
+      ));
+
+      return deleteResponse;
+    } catch (deleteError) {
+      return setErrorMessage('Unable to delete a todo');
+    }
   }, []);
 
+  const clearCompletedTodos = useCallback(() => {
+    todos.forEach(todo => {
+      if (todo.completed) {
+        deleteTodo(todo.id);
+      }
+    });
+  }, [todos]);
+
+  const addTodo = useCallback(
+    async (event: React.FormEvent<HTMLFormElement>, title: string) => {
+      event.preventDefault();
+
+      try {
+        if (!title.trim()) {
+          setErrorMessage('Title can\'t be empty');
+
+          return;
+        }
+
+        setIsNewTodoLoading(true);
+
+        if (user) {
+          setTempTodo({
+            id: 0,
+            userId: user.id,
+            title,
+            completed: false,
+          });
+
+          const newTodo = await createTodo({
+            userId: user.id,
+            title,
+            completed: false,
+          });
+
+          setTodos(prevTodos => [...prevTodos, newTodo]);
+          setTempTodo(null);
+          setIsNewTodoLoading(false);
+        }
+      } catch (addTodoError) {
+        setErrorMessage('Unable to add a todo');
```

Comment:
If you throw an error, the loader is still visible.
You should move this code inside `finally` instruction
```
setIsNewTodoLoading(false);
```

Reply:
fixed thank you